docs/ingressroute.md

```diff
@@ -0,0 +1,297 @@
+
+### Virtual Host Configuration
+
+### FQDN
+
+Similar to Ingress, IngressRoutes support name-based virtual hosting.  Name-based virtual hosts use multiple host names with the same IP address.
+
+```
+foo.bar.com --|                 |-> foo.bar.com s1:80
+              | 178.91.123.132  |
+bar.foo.com --|                 |-> bar.foo.com s2:80
+```
+
+Unlike, Ingress, however, IngressRoutes only support a single root domain per IngressRoute object.  As an example, this Ingress object:
+
+```
+# ingress-name.yaml
+apiVersion: extensions/v1beta1
+kind: Ingress
+metadata:
+  name: name-example
+spec:
+  rules:
+  - host: foo1.bar.com
+    http:
+      paths:
+      - backend:
+          serviceName: s1
+          servicePort: 80
+  - host: bar1.bar.com
+    http:
+      paths:
+      - backend:
+          serviceName: s2
+          servicePort: 80
+```
+
+must be represented by two different IngressRoute objects:
+```
+# ingressroute-name.yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: name-example-foo
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: foo1.bar.com
+  routes: 
+    - match: /
+      services: 
+        - name: s1
+          port: 80
+---
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: name-example-bar
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: bar1.bar.com
+  routes: 
+    - match: /
+      services: 
+        - name: s2
+          port: 80
+```
+
+### Aliases
+
+A single IngressRoute can define the routing configuration for a root domain and zero or more domain aliases.  This allows for sharing of configuration across multiple domains (e.g. `foo.com`, `www.foo.com`, and `bar.com`).
+
+```
+# domain-aliases.ingressroute.yaml
+apiVersion: contour.heptio.com/v1beta1
+kind: IngressRoute
+metadata: 
+  name: domain-aliases
+  namespace: default
+spec: 
+  virtualhost:
+    fqdn: bar.com
+    aliases:
+      - www.bar.com
+  routes: 
+    - match: /
+      services: 
+        - name: s1
+          port: 80
+```
+
+### TLS
+
+IngressRoutes follow a similar pattern for configuring TLS credentials as Ingress does.
+
+You can secure an IngressRoute by specifying a secret that contains a TLS private key and certificate. Currently, IngressRoutes only support a single TLS port, 443, and assume TLS termination.  If the `virtualhost` section includes domain aliases, the certificate must include the necessary Subject Authority Name (SAN) for each alias.  Contour (via Envoy) uses the SNI TLS extension to handle this behavior.
+
+The TLS secret must contain keys named tls.crt and tls.key that contain the certificate and private key to use for TLS, e.g.:
+
+```
+# ingress-tls.secret.yaml
+apiVersion: v1
+data:
+  tls.crt: base64 encoded cert
+  tls.key: base64 encoded key
+kind: Secret
+metadata:
+  name: testsecret
+  namespace: default
+type: Opaque
```

Comment:
We should use the `kubernetes.io/tls` type for TLS secrets: https://github.com/kubernetes/api/blob/master/core/v1/types.go#L4944